core/src/main/java/org/mobilitydata/gtfsvalidator/io/ValidationReportDeserializer.java

```diff
@@ -0,0 +1,114 @@
+
+/**
+ * Used to deserialize a JSON validation report. This represents a validation report as a list of
+ * {@code NoticeReport} which provides information about each notice generated during a GTFS dataset
+ * validation.
+ */
+public class ValidationReportDeserializer implements JsonDeserializer<ValidationReport> {
+
+  private static final Gson GSON =
+      new GsonBuilder().serializeNulls().serializeSpecialFloatingPointValues().create();
+  private static final String NOTICES_MEMBER_NAME = "notices";
+
+  /**
+   * Return the sorted set of error codes from a list of {@code NoticeReport}.
+   *
+   * @return the sorted set of error codes from a list of {@code NoticeReport}.
+   */
+  private static ImmutableSet<String> extractErrorCodes(Set<NoticeReport> notices) {
+    ImmutableSet.Builder<String> errorCodesSetBuilder = new ImmutableSet.Builder<>();
+    for (NoticeReport noticeReport : notices) {
+      if (noticeReport.isError()) {
+        errorCodesSetBuilder.add(noticeReport.getCode());
+      }
+    }
+    return errorCodesSetBuilder.build();
+  }
+
+  @Override
+  public ValidationReport deserialize(
+      JsonElement json, Type typoOfT, JsonDeserializationContext context) {
+    Set<NoticeReport> notices = new LinkedHashSet<>();
+    JsonObject rootObject = json.getAsJsonObject();
+    JsonArray noticesArray = rootObject.getAsJsonArray(NOTICES_MEMBER_NAME);
+    for (JsonElement childObject : noticesArray) {
+      notices.add(GSON.fromJson(childObject, NoticeReport.class));
+    }
+    return new ValidationReport(Collections.unmodifiableSet(notices), extractErrorCodes(notices));
+  }
+
+  public static <T extends Notice> ValidationReport serialize(
+      List<T> notices,
+      int maxExportsPerNoticeTypeAndSeverity,
+      Map<String, Integer> noticesCountPerTypeAndSeverity) {
+    Set<NoticeReport> noticeReports = new LinkedHashSet<>();
+    Gson gson = new Gson();
+    Set<String> errorCodes = new TreeSet<>();
+    Type contextType = new TypeToken<Map<String, Object>>() {}.getType();
+    for (Collection<T> noticesOfType :
+        NoticeContainer.groupNoticesByTypeAndSeverity(notices).asMap().values()) {
+      T firstNotice = noticesOfType.iterator().next();
+      if (firstNotice.isError()) {
+        errorCodes.add(firstNotice.getCode());
+      }
+      List<LinkedTreeMap<String, Object>> contexts = new ArrayList<>();
+      int i = 0;
+      for (T notice : noticesOfType) {
+        ++i;
+        if (i > maxExportsPerNoticeTypeAndSeverity) {
+          // Do not export too many notices for this type.
+          break;
+        }
+        contexts.add(gson.fromJson(notice.getContext(), contextType));
+      }
+      noticeReports.add(
+          new NoticeReport(
+              firstNotice.getCode(),
+              firstNotice.getSeverityLevel(),
+              noticesCountPerTypeAndSeverity.get(firstNotice.getMappingKey()),
+              contexts));
+    }
+    return new ValidationReport(noticeReports, errorCodes);
```

Comment:
`ValidationReport` has been refactored so that the extraction of `errorCodes` is done within the context of `ValidationReport` (no longer `ValidationReportDeserializer`): the constructor uses `Collections.unmodifiableSet`.